src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -808,11 +809,13 @@ NOINLINE freeze_result FreezeBase::recurse_freeze(frame& f, frame& caller, int c
     return recurse_freeze_compiled_frame(f, caller, callee_argsize, callee_interpreted);
   } else if (f.is_interpreted_frame()) {
     assert((_preempt && top) || !f.interpreter_frame_method()->is_native(), "");
-    if (_preempt && top && f.interpreter_frame_method()->is_native()) {
-      // int native entry
-      return freeze_pinned_native;
+    if (top) {
+      if (_preempt && f.interpreter_frame_method()->is_native()) {
+        // int native entry
+        return freeze_pinned_native;
+      }
+      prepare_freeze_interpreted_top_frame(f);
```

Comment:
Probably better to handle setting last_sp for this special case in freeze_start_frame_safepoint_stub(). But in any case the preemption code is not there yet so I would just remove these changes and prepare_freeze_interpreted_top_frame().